## Real-Time Mask Detection 

In [2]:
import os
import cv2
import pandas as pd
import numpy as np
from sklearn import preprocessing
from sklearn.model_selection import train_test_split 

In [3]:
Dir_W_M = []
Dir_WO_M = []
Dirs = []
Label = []


In [4]:
path_with_mask = "with_mask"
path_without_mask = "without_mask"

In [5]:
for r, d, f in os.walk(path_with_mask):
    for file in f:
        Dir_W_M.append(os.path.join(r, file))
        
for r, d, f in os.walk(path_without_mask):
    for file in f:
        Dir_WO_M.append(os.path.join(r, file))

## Preprocessing Steps:

In [6]:
for file in Dir_W_M:
    img = cv2.imread(file)
    gray = cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)
    resize = cv2.resize(gray,(200,200))
    Dirs.append(resize/255)
    Label.append('with_mask')
for file in Dir_WO_M:
    img = cv2.imread(file)
    gray = cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)
    resize = cv2.resize(gray,(200,200))
    Dirs.append(resize/255)
    Label.append('without_mask')

In [7]:
data = np.array(Dirs)
label_encoder = preprocessing.LabelEncoder() 
target = label_encoder.fit_transform(Label)

In [8]:
len(data)

930

In [9]:
data[0]

array([[0.22352941, 0.18823529, 0.19607843, ..., 0.0745098 , 0.15294118,
        0.1372549 ],
       [0.20392157, 0.18823529, 0.20392157, ..., 0.18431373, 0.10588235,
        0.15294118],
       [0.18823529, 0.2       , 0.21176471, ..., 0.25490196, 0.08235294,
        0.17254902],
       ...,
       [0.25098039, 0.24705882, 0.25098039, ..., 0.05882353, 0.03921569,
        0.03921569],
       [0.24705882, 0.23921569, 0.23921569, ..., 0.05098039, 0.05490196,
        0.05490196],
       [0.23137255, 0.22352941, 0.21960784, ..., 0.0627451 , 0.05882353,
        0.05882353]])

In [10]:
data[0].shape

(200, 200)

In [11]:
new_data = []
for i in data:
    new_data.append(i.reshape(-1))
    

In [12]:
len(new_data)

930

In [13]:
df = pd.DataFrame(new_data)
df

0         1         2         3         4         5         6      \
0    0.223529  0.188235  0.196078  0.200000  0.235294  0.247059  0.219608   
1    0.980392  0.980392  0.980392  0.984314  0.984314  0.984314  0.984314   
2    0.874510  0.874510  0.878431  0.882353  0.886275  0.890196  0.890196   
3    1.000000  1.000000  1.000000  1.000000  1.000000  1.000000  1.000000   
4    0.988235  0.968627  0.996078  0.960784  0.984314  1.000000  1.000000   
..        ...       ...       ...       ...       ...       ...       ...   
925  0.984314  0.980392  0.988235  0.988235  0.992157  0.988235  0.992157   
926  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
927  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
928  0.850980  0.878431  0.901961  0.968627  0.933333  0.949020  0.984314   
929  0.945098  0.933333  0.976471  0.960784  0.964706  0.941176  0.921569   

        7         8         9      ...     39990     39991     39992  \
0    0.172549  0.192157  0.180392  ...  0.090196  0.098039  0.133333   
1    0.984314  0.984314  0.984314  ...  0.086275  0.090196  0.090196   
2    0.898039  0.898039  0.898039  ...  0.156863  0.145098  0.137255   
3    1.000000  1.000000  1.000000  ...  1.000000  1.000000  1.000000   
4    0.996078  0.992157  0.992157  ...  0.352941  0.364706  0.305882   
..        ...       ...       ...  ...       ...       ...       ...   
925  0.988235  0.980392  0.984314  ...  0.498039  0.564706  0.698039   
926  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
927  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
928  0.988235  0.956863  0.956863  ...  0.113725  0.125490  0.133333   
929  0.933333  0.956863  0.886275  ...  0.070588  0.101961  0.109804   

        39993     39994     39995     39996     39997     39998     39999  
0    0.133333  0.047059  0.047059  0.054902  0.062745  0.058824  0.058824  
1    0.090196  0.086275  0.078431  0.078431  0.090196  0.090196  0.082353  
2    0.141176  0.149020  0.137255  0.117647  0.101961  0.094118  0.105882  
3    1.000000  1.000000  1.000000  1.000000  1.000000  1.000000  1.000000  
4    0.301961  0.305882  0.717647  0.741176  0.772549  0.227451  0.235294  
..        ...       ...       ...       ...       ...       ...       ...  
925  0.898039  0.976471  0.890196  0.215686  0.015686  0.023529  0.556863  
926  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  
927  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  
928  0.090196  0.047059  0.086275  0.160784  0.160784  0.658824  0.909804  
929  0.113725  0.176471  0.133333  0.164706  0.156863  0.717647  0.874510  

[930 rows x 40000 columns]

In [14]:
print(new_data[0])

[0.22352941 0.18823529 0.19607843 ... 0.0627451  0.05882353 0.05882353]


In [15]:
new_data[0].shape

(40000,)

In [16]:
X_train, X_test, y_train, y_test = train_test_split(df, target, test_size=0.20)

In [17]:
len(X_train)

744

In [18]:
len(X_test)

186

## Random Forest Classifier

In [19]:
from sklearn.metrics import confusion_matrix

In [20]:
from sklearn.ensemble import RandomForestClassifier

Random_forest = RandomForestClassifier(criterion='entropy')
Random_forest.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='entropy', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [21]:
Random_forest.score(X_test,y_test)

0.989247311827957

In [22]:
y_pred = Random_forest.predict(X_test)
confusion_matrix(y_test, y_pred)

array([[153,   0],
       [  2,  31]], dtype=int64)

### Accuracy
(TP + TN) / (TP+FP+FN+TN) = (130+52)/ (186) = 0.962962

### Precision
TP/ (TP+FP) = 130/ 131 = 0.9923664122

### Recall
TP / (TP+FN) = 130 / 133 = 0.977443609

### Specificity
TN / (TN+FP) = 52 / 53 = 0.9811320755

### F1 - SCORE
2(PRECISION * RECALL)/ (PRECISION + RECALL) = 2( 0.992 0.977) / (0.992 + 0.977) = 0.9844428644

## K-Neighbors Classifier

In [23]:
from sklearn.neighbors import KNeighborsClassifier

KNN_model = KNeighborsClassifier(n_neighbors=27)

KNN_model.fit(X_train,y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=27, p=2,
                     weights='uniform')

In [24]:
KNN_model.score(X_test,y_test)

0.8924731182795699

In [25]:
y_pred = KNN_model.predict(X_test)
confusion_matrix(y_test, y_pred)

array([[150,   3],
       [ 17,  16]], dtype=int64)

### Accuracy
(TP + TN) / (TP+FP+FN+TN) = (121 + 43)/ (186) = 0.8817204301

### Precision
TP/ (TP+FP) = 121/ 131 = 0.9236641221

### Recall
TP / (TP+FN) = 121 / 133 = 0.9097744361

### Specificity
TN / (TN+FP) = 43 / 53 = 0.8113207547

### F1 - SCORE
2(PRECISION * RECALL)/ (PRECISION + RECALL) = 2( 0.923 * 0.909) / ( 0.923 + 0.909) = 0.9159465066

##  Support Vector Machine

In [26]:
from sklearn.svm import SVC
svclassifier = SVC(kernel='linear')
svclassifier.fit(X_train, y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='linear',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [27]:
svclassifier.score(X_test,y_test)

0.9408602150537635

In [28]:
y_pred = svclassifier.predict(X_test)
confusion_matrix(y_test, y_pred)

array([[144,   9],
       [  2,  31]], dtype=int64)

### Accuracy
(TP + TN) / (TP+FP+FN+TN) = (123 + 48)/ (186) = 0.9193548387

### Precision
TP/ (TP+FP) = 123/ 131 = 0.9389312977

### Recall
TP / (TP+FN) = 123 / 130 = 0.9461538462

### Specificity
TN / (TN+FP) = 48 / 56 = 0.8571428571

### F1 - SCORE
2(PRECISION * RECALL)/ (PRECISION + RECALL) = 2( 0.938 * 0.946) / ( 0.938 + 0.946) = 0.9419830149

## Logistic Regression

In [29]:
from sklearn.linear_model import LogisticRegression
Logestic = LogisticRegression(random_state=0).fit(X_train, y_train)

C:\Users\OVIYA H\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [30]:
Logestic.score(X_test,y_test)

0.9301075268817204

In [31]:
y_pred = Logestic.predict(X_test)
confusion_matrix(y_test, y_pred)

array([[142,  11],
       [  2,  31]], dtype=int64)

### Accuracy
(TP + TN) / (TP+FP+FN+TN) = (125 + 42)/ (186) = 0.8978494623

### Precision
TP/ (TP+FP) = 125 / 131 = 0.9541984733

### Recall
TP / (TP+FN) = 125 / 138 = 0.9057971014

### Specificity
TN / (TN+FP) = 42 / 48 = 0.875

### F1 - SCORE
2(PRECISION * RECALL)/ (PRECISION + RECALL) = 2( 0.954 * 0.905) / ( 0.954 + 0.905) = 0.9288542227

## Real time Model Prediction

In [32]:
Source = cv2.VideoCapture(0) 
  
while(True): 
    ret, frame = Source.read() 
    face_cascade = cv2.CascadeClassifier('haarcascade_frontalface_default.xml')

    faces = face_cascade.detectMultiScale(frame)
    
    for (x,y,w,h) in faces:
        crop_img = frame[y:y+h, x:x+w]
        gray = cv2.cvtColor(crop_img,cv2.COLOR_BGR2GRAY)  
        resize = cv2.resize(gray,(200,200))

        resize = resize/255
        data = np.array(resize)
        
        data = data.reshape(-1)
        result = svclassifier.predict([data])
        print(result)
        
        if result == 0:
            cv2.rectangle(frame, (x, y), (x + w, y + h),(0,255,0), 2) 
            cv2.putText(frame, 'With MASK', (x, y-10), cv2.FONT_HERSHEY_SIMPLEX, 0.9, (255,255,255), 2)
        else:
            cv2.rectangle(frame, (x, y), (x + w, y + h),(0,0,255), 2) 
            cv2.putText(frame, 'Without mask', (x, y-10), cv2.FONT_HERSHEY_SIMPLEX, 0.9, (255,255,255), 2)
            
        
    
    cv2.imshow('img',frame)
    
    if cv2.waitKey(1) & 0xFF == ord('q'): 
        break

  
Source.release() 
cv2.destroyAllWindows() 

[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[0]
[0]
[0]
[0]
[0]
[1]
[0]
[1]
[0]
[0]
[1]
[0]
[0]
[0]
[0]
[0]
[0]
[1]
[0]
[1]
[1]
[1]
[1]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[1]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[0]
[1]
[1]
[0]
[0]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
[1]
